_episodes/32.md

```diff
@@ -0,0 +1,389 @@
+</p>
+
+Buffering intermediate results makes it possible to execute queries iteratively.
+For example, the engine can process one or several tasks at a time effectively 
+reducing memory pressure and allow memory intensive queries to succeed without 
+a need to expand the cluster. It would significantly lower cost of operation, 
+specifically for the situation when only a small number of queries requires more
+memory than available.
+
+<p align="center">
+<img align="center" width="50%" height="100%" src="/assets/episode/32/tardigrade-buffers.png"/><br/>
+</p>
+
+#### Adaptive Planning
+
+The engine may also decide to re-optimize the query at stage boundaries. When  
+the engine buffers the intermediate data, it is possible to get better insight
+into the nature of the data as it's processed and adapt query plans accordingly.
+For example, when the cost based optimizer makes a bad decision, because of 
+incorrect statistics or estimates, it can pick the wrong type of join, or a 
+suboptimal join order. The engine can then suspend the query, re-optimize the 
+plan, and resume the query. Additionally, it may allow the engine to discover 
+skewed datasets and change query plans accordingly. This may significantly 
+improve efficiency and landing time for workloads that are `JOIN` heavy.
+
+#### Resource Management
+
+Iterative query processing would also allow us to be more flexible at resource 
+management. Resource allocation can be adjusted as the queries run. For example,
+when a cluster is idle we may allow a single query to utilize all available 
+resources on a cluster. When more workload kicks in the resource allocation for
+the initial query can be gradually reduced and available resources can be 
+granted to newly submitted workloads. With this model it is also significantly 
+easier to implement auto scaling. When workload submitted requires more 
+resources that are currently available in the cluster the engine may request 
+more nodes to be added to the cluster. Or quite the opposite, if the cluster 
+is being underutilized it is significantly easier to return resources when 
+there’s no need to wait for slow running tasks. Being able to better manage 
+available resources and adjust the resource pool based on the current workload 
+submitted would make the engine significantly more cost effective.
+
+#### Fine-grained failure recovery
+
+Last, but not least, with project Tardigrade we are going to provide 
+fine-grained failure recovery. The buffering introduced at stage boundaries 
+would allow a transparent restart of failed tasks. Fine grained failure recovery
+would make landing time for ETL pipelines significantly more predictable. Also,
+it opens the opportunity of running ETL workload on much cheaper widely 
+available spot instances that can further optimize operational costs.
+
+### Opportunities that Tardigrade opens
+
+In summary, Project Tardigrade will offer the following two categories of
```

Comment:
In summary, in Project Tardigrade we work on the following improvements of Trino: